Terminal.Gui/Views/FileDialog.cs

```diff
@@ -1,1053 +1,1634 @@
-		/// <param name="nameFieldLabel">The name of the file field label..</param>
-		/// <param name="message">The message.</param>
-		/// <param name="allowedTypes">The allowed types.</param>
-		public FileDialog (ustring title, ustring prompt, ustring nameFieldLabel, ustring message, List<string> allowedTypes = null)
-			: this (title, prompt, ustring.Empty, nameFieldLabel, message, allowedTypes) { }
+		/// <remarks><see cref="AllowedTypeAny"/> adds the option to select any type (*.*). If this
+		/// collection is empty then any type is supported and no Types drop-down is shown.</remarks> 
+		public List<IAllowedType> AllowedTypes { get; set; } = new List<IAllowedType> ();
 
 		/// <summary>
-		/// Initializes a new instance of <see cref="FileDialog"/>
+		/// Gets a value indicating whether the <see cref="FileDialog"/> was closed
+		/// without confirming a selection.
 		/// </summary>
-		/// <param name="title">The title.</param>
-		/// <param name="prompt">The prompt.</param>
-		/// <param name="message">The message.</param>
-		/// <param name="allowedTypes">The allowed types.</param>
-		public FileDialog (ustring title, ustring prompt, ustring message, List<string> allowedTypes)
-			: this (title, prompt, ustring.Empty, message, allowedTypes) { }
+		public bool Canceled { get; private set; } = true;
 
 		/// <summary>
-		/// Initializes a new instance of <see cref="FileDialog"/>
+		/// Gets all files/directories selected or an empty collection
+		/// <see cref="AllowsMultipleSelection"/> is <see langword="false"/> or <see cref="Canceled"/>.
 		/// </summary>
-		/// <param name="title">The title.</param>
-		/// <param name="prompt">The prompt.</param>
-		/// <param name="nameDirLabel">The name of the directory field label.</param>
-		/// <param name="nameFieldLabel">The name of the file field label..</param>
-		/// <param name="message">The message.</param>
-		/// <param name="allowedTypes">The allowed types.</param>
-		public FileDialog (ustring title, ustring prompt, ustring nameDirLabel, ustring nameFieldLabel, ustring message,
-			List<string> allowedTypes = null) : base (title)//, Driver.Cols - 20, Driver.Rows - 5, null)
+		/// <remarks>If selecting only a single file/directory then you should use <see cref="Path"/> instead.</remarks>
+		public IReadOnlyList<string> MultiSelected { get; private set; }
+			= Enumerable.Empty<string> ().ToList ().AsReadOnly ();
+
+
+		/// <inheritdoc/>
+		public override void Redraw (Rect bounds)
 		{
-			this.message = new Label (message) {
-				X = 1,
-				Y = 0,
-			};
-			Add (this.message);
-			var msgLines = TextFormatter.MaxLines (message, Driver.Cols - 20);
+			base.Redraw (bounds);
 
-			this.nameDirLabel = new Label (nameDirLabel.IsEmpty ? $"{Strings.fdDirectory}: " : $"{nameDirLabel}: ") {
-				X = 1,
-				Y = 1 + msgLines,
-				AutoSize = true
-			};
+			if (!string.IsNullOrWhiteSpace (feedback)) {
+				var feedbackWidth = feedback.Sum (c => Rune.ColumnWidth (c));
```

Comment:
Can you use `View.Margin` for this instead?

Reply:
Sure, is there docs on this?  I tried

```csharp
lblFeedback = new Label ("Test") {
		Y = Pos.AnchorEnd(1),
		Width = Dim.Fill()
};
Margin.Thickness = new Thickness(0,0,0,1);
Margin.Add(lblFeedback);
```

Reply:
Actually @tig I'm not sure I understand the change you are looking for here with Margin.  This is the current user interaction.  

- Error messages appear in center of Dialog
- Any keypress in the Dialog clears the message

If it were in Margin wouldn't it have to be at the top/bottom? Please are you able to provide some pseudocode for what you were thinking of?

I'm open to suggestions on how to improve UX.  I only did it this way to avoid throwing up a modal popup over another popup (e.g. an error popup).  But didn't want to give the user no feedback at all that they were trying to take an action that was being prevented.

![file-must-exist](https://user-images.githubusercontent.com/31306100/230762392-b802884e-f41e-47c0-a86f-db91aa3fbb66.gif)

Reply:
> Sure, is there docs on this? My first instinct I tried was variations on:
> 
> ```cs
> lblFeedback = new Label ("Test") {
> 		Y = Pos.AnchorEnd(1),
> 		Width = Dim.Fill()
> };
> Margin.Thickness = new Thickness(0,0,0,1);
> Margin.Add(lblFeedback);
> ```

Adding views to `Frame`s is not yet supported. 

What I was asking about is the logic you have for calculating spacing between the subviews.  

Instead of all of that, it might be possible to just set their `Margin.Thickness` to values that get you the spacing you want.

Reply:
This is no big deal; and I think I may have misread your code.